### PR TITLE
feat: Copy a Zarr archive to the Hub using copy_store for increased efficient

### PR DESCRIPTION
**Pull request by lmtroper**
## Changelogs

Replaced `copy_all` with `copy_store` when copying a Zarr archive to the Hub. The changes should increase efficiency as `copy_store` does not de-compress and re-compress the data during copying ([see documentation here](https://zarr.readthedocs.io/en/stable/api/convenience.html#zarr.convenience.copy_store)).

Below is the impact it had on profiling:

With `copy_all`:
```Profiling Report
====================================================================================================
Date: 2024-04-04
Time: 17:29:52
Size: 953.67 MB
Repeats: 5
Polaris version: 0.0.2.dev191+g82e7db2
Zarr version: 2.17.1
====================================================================================================
                         Creating the Zarr archive: 0:00:01.344764 ± 0:00:00.151870
         Creating dataset from Source Zarr archive: 0:00:01.509610 ± 0:00:00.201476
                      Uploading dataset to the Hub: 0:08:58.749664 ± 0:00:24.787134
                          Loading dataset from Hub: 0:00:02.294858 ± 0:00:00.129297
                          Caching dataset to local: 0:10:36.877163 ± 0:00:35.494129
                    Iterating over dataset (local): 2:05:54.817060 ± 0:02:11.140053
           Baseline Zarr only upload to Cloudflare: 0:00:30.342002 ± 0:00:01.761582
             Baseline dataset upload to Cloudflare: 0:00:34.035674 ± 0:00:00.615563
       Baseline Zarr only download from Cloudflare: 0:00:44.472856 ± 0:00:21.117384
         Baseline dataset download from Cloudflare: 0:00:33.714192 ± 0:00:01.377525
====================================================================================================
                        Actual / Baseline - Upload: 15.843 ± 0.939
                      Actual / Baseline - Download: 18.926 ± 1.369
```

With `copy_store`:

```Profiling Report
====================================================================================================
Date: 2024-04-09
Time: 06:45:36
Size: 953.67 MB
Repeats: 1
Polaris version: dev
Zarr version: 2.16.1
====================================================================================================
                         Creating the Zarr archive: 0:00:01.407685 ± 0:00:00
         Creating dataset from Source Zarr archive: 0:00:01.874611 ± 0:00:00
                      Uploading dataset to the Hub: 0:28:05.252655 ± 0:00:00
                          Loading dataset from Hub: 0:00:03.237522 ± 0:00:00
                          Caching dataset to local: 0:16:53.128378 ± 0:00:00
           Baseline Zarr only upload to Cloudflare: 0:04:31.668470 ± 0:00:00
             Baseline dataset upload to Cloudflare: 0:04:43.972480 ± 0:00:00
       Baseline Zarr only download from Cloudflare: 0:01:09.703651 ± 0:00:00
         Baseline dataset download from Cloudflare: 0:01:11.038993 ± 0:00:00
====================================================================================================
                        Actual / Baseline - Upload: 5.935 ± 0.000
                      Actual / Baseline - Download: 14.262 ± 0.000
```

**Note:** We did not implement `copy_store` for downloading datasets (only uploading) as it resulted in an increase in caching the dataset locally. See below:

`copy_store` for both upload and download:

```Profiling Report
====================================================================================================
Date: 2024-04-09
Time: 14:02:09
Size: 95.37 MB
Repeats: 1
Polaris version: dev
Zarr version: 2.16.1
====================================================================================================
                         Creating the Zarr archive: 0:00:00.223971 ± 0:00:00
         Creating dataset from Source Zarr archive: 0:00:00.242384 ± 0:00:00
                      Uploading dataset to the Hub: 0:07:06.433739 ± 0:00:00
                          Loading dataset from Hub: 0:00:02.718317 ± 0:00:00
                          Caching dataset to local: 0:04:54.390313 ± 0:00:00
           Baseline Zarr only upload to Cloudflare: 0:00:32.339057 ± 0:00:00
             Baseline dataset upload to Cloudflare: 0:00:33.148623 ± 0:00:00
       Baseline Zarr only download from Cloudflare: 0:00:09.478066 ± 0:00:00
         Baseline dataset download from Cloudflare: 0:00:07.696965 ± 0:00:00
====================================================================================================
                        Actual / Baseline - Upload: 12.864 ± 0.000
                      Actual / Baseline - Download: 38.248 ± 0.000
```
                      
                      
                      
`copy_store` for just upload:

```Profiling Report
====================================================================================================
Date: 2024-04-09
Time: 14:16:02
Size: 95.37 MB
Repeats: 1
Polaris version: dev
Zarr version: 2.16.1
====================================================================================================
                         Creating the Zarr archive: 0:00:00.188619 ± 0:00:00
         Creating dataset from Source Zarr archive: 0:00:00.235259 ± 0:00:00
                      Uploading dataset to the Hub: 0:06:52.969075 ± 0:00:00
                          Loading dataset from Hub: 0:00:02.182547 ± 0:00:00
                          Caching dataset to local: 0:04:19.006523 ± 0:00:00
           Baseline Zarr only upload to Cloudflare: 0:00:32.249822 ± 0:00:00
             Baseline dataset upload to Cloudflare: 0:00:33.095662 ± 0:00:00
       Baseline Zarr only download from Cloudflare: 0:00:07.731580 ± 0:00:00
         Baseline dataset download from Cloudflare: 0:00:07.948864 ± 0:00:00
====================================================================================================
                        Actual / Baseline - Upload: 12.478 ± 0.000
                      Actual / Baseline - Download: 32.584 ± 0.000

### CODE REVIEW

File: polaris/hub/client.py [hunk 1]
```diff
@@ -617,8 +617,8 @@ def upload_dataset(
                 dest.store[".zmetadata"] = zmetadata_content
 
                 logger.info("Copying Zarr archive to the Hub. This may take a while.")
-                zarr.copy_all(source=dataset.zarr_root, dest=dest, log=logger.info)
-
+                zarr.copy_store(source=dataset.zarr_root.store.store, dest=dest.store, log=logger.info, if_exists='replace')
```
Comment (cwognum) [inline polaris/hub/client.py:620]: I'm not sure about setting the `if_exists` parameter here. I think all values (replace, skip, raise) can be useful. For example, if you accidentally stop your download, you'll likely want to use "skip" to continue where you left off. 

Maybe add it as a parameter to the `upload_dataset method` and keep it as "raise" by default?

**Review by cwognum — CHANGES_REQUESTED** [review]
Thanks @lmtroper! I believe we can we use the same on download! It should speed things up further!

File: polaris/hub/client.py [hunk 2]
```diff
@@ -497,6 +504,7 @@ def upload_dataset(
         access: AccessType = "private",
         timeout: TimeoutTypes = (10, 200),
         owner: Optional[Union[HubOwner, str]] = None,
+        on_conflict: ZarrConflictResolution = "raise",
```
Comment (cwognum) [inline polaris/hub/client.py:507]: To minimize friction for people that are familiar with Zarr, I suggest we keep the name the same as for Zarr: `if_exists`.

File: polaris/utils/types.py [hunk 3]
```diff
@@ -99,6 +99,11 @@
 Supported license types for dataset uploads to Polaris Hub
 """
 
+ZarrConflictResolution: TypeAlias = Literal["raise", "replace", "skip"]
```
Comment (cwognum) [inline polaris/utils/types.py:102]: Introducing a new type is a good call here! Thanks! :pray:

**Comment by cwognum** [comment]
@lmtroper Can you replace [`zarr.copy_all` in the download method](https://github.com/polaris-hub/polaris/blob/main/polaris/dataset/_dataset.py#L352) as well?

File: polaris/hub/client.py (hunk 1 above)
Comment (lmtroper) [reply polaris/hub/client.py]: @cwognum Yesterday I noticed that during upload, I was adding (i.e., replacing in this case) .zgroup to the same path twice. Just raising this because I would then assume that most, if not all, archives on initial upload might hit this error.

Comment (cwognum) [reply polaris/hub/client.py]: Hm, good catch! Maybe that it's because when we open the Zarr archive, it already creates some files?

Maybe we shouldn't open the archive, but just create the respective store object. Could you try replacing [this line of code](https://github.com/polaris-hub/polaris/blob/main/polaris/dataset/_dataset.py#L351):
```py
dest = zarr.open(zarr_archive, "w")
```

by something like: 
```py
dest = zarr.DirectoryStore(zarr_archive)
```

(similar for the upload process)

File: polaris/dataset/_dataset.py [hunk 4]
```diff
@@ -348,14 +348,15 @@ def to_json(self, destination: str) -> str:
 
         # Copy over Zarr data to the destination
         if self.zarr_root is not None:
-            dest = zarr.open(zarr_archive, "w")
-            zarr.copy_all(source=self.zarr_root, dest=dest)
+            dest = zarr.DirectoryStore(zarr_archive)
+
+            zarr.copy_store(source=self.zarr_root.store.store, dest=dest)
```
Comment (cwognum) [inline polaris/dataset/_dataset.py:353]: Could you try see if this works?

 ```suggestion
             zarr.copy_store(source=self.zarr_root.store, dest=dest)
```

I think it would, but maybe not!

File: polaris/hub/client.py [hunk 5]
```diff
@@ -527,15 +527,15 @@ def upload_dataset(
...
 
         # Check if a dataset license was specified prior to upload
-        if not dataset.license:
-            raise InvalidDatasetError(
-                f"\nPlease specify a supported license for this dataset prior to uploading to the Polaris Hub.\nOnly some Creative Commons licenses are supported - {get_args(SupportedLicenseType)}. For more information, see https://creativecommons.org/share-your-work/cclicenses/"
-            )
+        # if not dataset.license:
+        #     raise InvalidDatasetError(
+        #         f"\nPlease specify a supported license for this dataset prior to uploading to the Polaris Hub.\nOnly some Creative Commons licenses are supported - {get_args(SupportedLicenseType)}. For more information, see https://creativecommons.org/share-your-work/cclicenses/"
+        #     )
```
Comment (cwognum) [inline polaris/hub/client.py]: This should be uncommented!

File: polaris/hub/client.py [hunk 6]
```diff
@@ -610,28 +610,24 @@ def upload_dataset(
...
                 # Locally consolidate Zarr archive metadata. Future updates on handling consolidated
                 # metadata based on Zarr developers' recommendations can be tracked at:
                 # https://github.com/zarr-developers/zarr-python/issues/1731
                 zarr.consolidate_metadata(dataset.zarr_root.store.store)
-                zmetadata_content = dataset.zarr_root.store.store[".zmetadata"]
-                dest.store[".zmetadata"] = zmetadata_content
 
                 logger.info("Copying Zarr archive to the Hub. This may take a while.")
                 zarr.copy_store(
                     source=dataset.zarr_root.store.store,
```
Comment (cwognum) [inline polaris/hub/client.py:627]: Similar question here: would this work?

```suggestion
                    source=dataset.zarr_root.store,
```

File: polaris/hub/client.py [hunk 7]
```diff
@@ -610,28 +610,24 @@ def upload_dataset(
...
+                    dataset_name=dataset.name,
                 )
+                dest = zarr.storage.FSStore(url=dataset_json["zarrRootPath"], fs=polaris_fs, mode="w")
 
                 # Locally consolidate Zarr archive metadata. Future updates on handling consolidated
                 # metadata based on Zarr developers' recommendations can be tracked at:
                 # https://github.com/zarr-developers/zarr-python/issues/1731
                 zarr.consolidate_metadata(dataset.zarr_root.store.store)
-                zmetadata_content = dataset.zarr_root.store.store[".zmetadata"]
-                dest.store[".zmetadata"] = zmetadata_content
```
Comment (cwognum) [inline polaris/hub/client.py:617]: Why was this deleted?? If we can delete it here, can we delete it in `to_json` too?

File: polaris/hub/client.py [hunk 8]
```diff
@@ -610,28 +610,24 @@ def upload_dataset(
...
+                    polaris_client=self,
+                    dataset_owner=dataset.owner,
+                    dataset_name=dataset.name,
                 )
+                dest = zarr.storage.FSStore(url=dataset_json["zarrRootPath"], fs=polaris_fs, mode="w")
 
                 # Locally consolidate Zarr archive metadata. Future updates on handling consolidated
                 # metadata based on Zarr developers' recommendations can be tracked at:
                 # https://github.com/zarr-developers/zarr-python/issues/1731
                 zarr.consolidate_metadata(dataset.zarr_root.store.store)
```
Comment (cwognum) [inline polaris/hub/client.py:623]: Since we by now assume a Zarr archive is always consolidated, we can remove these lines, right?

File: polaris/dataset/_dataset.py (hunk 4 above)
Comment (lmtroper) [reply polaris/dataset/_dataset.py:353]: Ya I previously tried this and run into the following error:

```Traceback (most recent call last):
  File "/Users/larissa.troper/polaris/profiling.py", line 350, in <module>
    app()
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/typer/main.py", line 328, in __call__
    raise e
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/typer/main.py", line 311, in __call__
    return get_command(self)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/typer/core.py", line 716, in main
    return _main(
           ^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/typer/core.py", line 216, in _main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/typer/main.py", line 683, in wrapper
    return callback(**use_params)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/polaris/profiling.py", line 296, in profile
    profile_all(tmpdir, owner, name, n_bytes, n_cols, save_dir, fs, report=report)
  File "/Users/larissa.troper/polaris/profiling.py", line 189, in profile_all
    dataset.cache(os.path.join(tmpdir, "cache"))
  File "/Users/larissa.troper/polaris/polaris/dataset/_dataset.py", line 383, in cache
    self.to_json(self.cache_dir)
  File "/Users/larissa.troper/polaris/polaris/dataset/_dataset.py", line 353, in to_json
    zarr.copy_store(source=self.zarr_root.store, dest=dest)
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/zarr/convenience.py", line 732, in copy_store
    n_bytes_copied += buffer_size(data)
                      ^^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/zarr/util.py", line 380, in buffer_size
    return ensure_ndarray_like(v).nbytes
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/larissa.troper/anaconda3/envs/polaris/lib/python3.12/site-packages/numcodecs/compat.py", line 42, in ensure_ndarray_like
    mem = memoryview(buf)
          ^^^^^^^^^^^^^^^
TypeError: memoryview: a bytes-like object is required, not 'dict'```

File: polaris/hub/client.py (hunk 7 above)
Comment (lmtroper) [reply polaris/hub/client.py:617]: The changes resulted in `zarr.errors.CopyError: key '.zmetadata' exists in destination` so I had to remove it. Should also be good to remove it from `to_json`.

File: polaris/hub/client.py (hunk 6 above)
Comment (lmtroper) [reply polaris/hub/client.py:627]: ^^ Same error as mentioned above.

File: polaris/hub/client.py (hunk 7 above)
Comment (cwognum) [reply polaris/hub/client.py:617]: Interesting! So `copy_store` solves https://github.com/zarr-developers/zarr-python/issues/1731?

**Review by cwognum — APPROVED** [review]
